File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        HashMap<Integer,Integer>m=new HashMap<>();
        for(int i=0;i<inorder.length;i++){
            m.put(inorder[i],i);
        }
        TreeNode root=built(preorder,m,0,inorder.length-1,0);
        return root;
    }
    public TreeNode built(int[] preorder,HashMap<Integer,Integer>m,int left,int right,int rootIndex){
        TreeNode root=new TreeNode(preorder[rootIndex]);
        int mid=m.get(preorder[rootIndex]);
        if(left<mid){
          root.left=built(preorder,m,left,mid-1,rootIndex+1);
        }
        if(right>mid){
        root.right=built(preorder,m,mid+1,right,rootIndex+mid-left+1);
     
        }


return root;


    }
}

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        mp={}
        for i in range(len(preorder)):
            mp[inorder[i]]=i
        root=self.built(mp,inorder,preorder,0,0,len(inorder)-1)
        return root
    def built(self,mp,inorder,preorder,rootindex,left,right):
        root=TreeNode(preorder[rootindex])
        mid=mp.get(preorder[rootindex])
        if left<mid:
            root.left=self.built(mp,inorder,preorder,rootindex+1,left,mid-1)
        if mid<right:
            root.right=self.built(mp,inorder,preorder,rootindex+1+mid-left,mid+1,right)
        return root